src/analyses/goto_check.cpp

```diff
@@ -1880,7 +1917,13 @@ void goto_checkt::goto_check(
     {
       const code_assignt &code_assign=to_code_assign(i.code);
 
+      // Question: I am not sure whether adding it as a class field is
+      // better than adding it as a function argument. Adding it as an
+      // argument means that it needs to be passed as a third argument
+      // to many check_rec functions that don't really need it.
+      no_enum_check = true;
```

Comment:
Not a fan of the class-field style here (it is likely to go wrong if exceptions are caught, or it may not nest well), but when can you get enum values being used on the LHS anyhow?

Reply:
If I remember correctly, enum type variables are on the LHS, and this assertion always fails when they are not initialized. 

Regarding the class-field, do you think it would be better passed as an argument to check_rec?

Reply:
Mmm, neither option is good is it... how about using an on-stack class instance that memorises the old state and stores it, then at least it's exception safe and nests properly, that's probably the least bad option

Reply:
I am not sure I understand your proposal. Could you clarify that? Thanks!

Reply:
For example,

```
class disable_no_enum_checkt
{
  disable_no_enum_checkt(goto_checkt &current_state) :
    goto_check(current_state),
    old_value(current_state.no_enum_check)
  { 
    goto_check.no_enum_check = true;
  }

  ~disable_no_enum_checkt()
  {
    goto_check.no_enum_check = old_value;
  }
};
```

Then
```
  {
    disable_no_enum_checkt disable_check(*this);
    process_lhs(...);
  }
```
The advantage being, (a) does the right thing in the presence of exceptions (i.e. restores the setting to its old value), and (b) notices if the check was already disabled (e.g. recursing into a function that disables it), rather than always re-enabling it on exit.

Reply:
@smowton Thanks for the suggestion! A minor comment is that this function cannot be recursively called, and so there shouldn't be any recursion issues. In that case, I guess that we can get away with:

```cpp
class disable_no_enum_checkt
{
  goto_checkt &goto_check;

public:
  disable_no_enum_checkt(goto_checkt &current_state) :
    goto_check(current_state)
  { 
    goto_check.no_enum_check = true;
  }

  ~disable_no_enum_checkt()
  {
    goto_check.no_enum_check = false;
  }
};
```

Would this work?

Reply:
Personally I'd still remember the old value to defend against this code being cut-and-pasted inside the recursion at some later point

Reply:
That is a reasonable point, thanks

Reply:
I just noticed that a similar `flag_resett` class exists, and I used that instead. @smowton it seems that it does exactly what you proposed.